lls/cache.c

```diff
@@ -0,0 +1,428 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+
+#include <rte_hash.h>
+#include <rte_malloc.h>
+
+#include "gatekeeper_net.h"
+#include "cache.h"
+
+#ifdef RTE_MACHINE_CPUFLAG_SSE4_2
+#include <rte_hash_crc.h>
+#define DEFAULT_HASH_FUNC       rte_hash_crc
+#else
+#include <rte_jhash.h>
+#define DEFAULT_HASH_FUNC       rte_jhash
+#endif
+
+/*
+ * Number of entries that a LLS cache can hold.
+ * This is also the number of entries stored in
+ * a mempool, which performs best when the size
+ * is a power of 2 minus 1.
+ */
+#define LLS_CACHE_ENTRIES		1023
+/*
+ * Size of mempool cache for LLS cache entries.
+ * This should be a divisor of LLS_CACHE_ENTRIES.
+ */
+#define LLS_ENTRIES_MP_CACHE_SIZE	77
+
+/* XXX Sample parameters, need to be tested for better performance. */
+#define LLS_CACHE_BURST_SIZE		32
+
+static void
+lls_cache_dump(struct lls_cache *cache)
+{
+	uint32_t iter = 0;
+	int32_t index;
+	const void *ip_key;
+	void *data;
+
+	if (unlikely(cache == NULL)) {
+		RTE_LOG(ERR, GATEKEEPER, "lls: %s: %s cache not instantiated\n",
+			__func__, cache->name);
+		return;
+	}
+
+	RTE_LOG(INFO, GATEKEEPER, "LLS cache (%s)\n=====================\n",
+		cache->name);
+	index = rte_hash_iterate(cache->hash, &ip_key, &data, &iter);
+	while (index >= 0) {
+		cache->print_entry(cache, ip_key, cache->data[index]);
+		index = rte_hash_iterate(cache->hash, &ip_key, &data, &iter);
+	}
+}
+
+static void
+__lls_cache_add(struct lls_cache *cache, const void *ip_key,
+	struct lls_cache_entry *new_entry)
+{
+	int32_t ret = rte_hash_add_key(cache->hash, ip_key);
+	if (unlikely(ret == -EINVAL)) {
+		RTE_LOG(ERR, HASH, "Invalid params, %s entry not added\n",
+			cache->name);
+		rte_mempool_put(cache->entry_pool, new_entry);
+	} else if (unlikely(ret == -ENOSPC)) {
+		RTE_LOG(ERR, HASH, "No space in cache, %s entry not added\n",
+			cache->name);
+		rte_mempool_put(cache->entry_pool, new_entry);
+	} else {
+		struct lls_cache_entry *existing_entry = cache->data[ret];
```

Comment:
In previous reviews, we removed some `RTE_ASSERT` because we wanted to fail gracefully in code after setup is completed. Following that, should we just add an error message here instead?